metricbeat/module/vsphere/host/host.go

```diff
@@ -240,3 +218,62 @@ func getAssetNames(ctx context.Context, pc *property.Collector, hs *mo.HostSyste
 		outputVmNames:      outputVmNames,
 	}, nil
 }
+
+func (m *HostMetricSet) getPerfMetrics(ctx context.Context, perfManager *performance.Manager, hst mo.HostSystem, metricIds []types.PerfMetricId) (map[string]interface{}, error) {
+	metricMap := make(map[string]interface{})
+	summary, err := perfManager.ProviderSummary(ctx, hst.Reference())
+	if err != nil {
+		return metricMap, fmt.Errorf("failed to get summary: %w", err)
+	}
+
+	var refreshRate = int32(m.Module().Config().Period.Seconds())
+	if summary.CurrentSupported {
+		refreshRate = summary.RefreshRate
+		if int32(m.Module().Config().Period.Seconds()) != refreshRate {
+			m.Logger().Warnf("User-provided period %v does not match system's refresh rate %v. Risk of data duplication. Consider adjusting period.", m.Module().Config().Period, refreshRate)
+		}
+	} else {
+		m.Logger().Warnf("Live data collection not supported. Use one of the system's historical interval (300, 1800, 7200, 86400). Risk of data duplication. Consider adjusting period.")
+	}
+
+	spec := types.PerfQuerySpec{
+		Entity:     hst.Reference(),
+		MetricId:   metricIds,
+		MaxSample:  1,
+		IntervalId: refreshRate,
+	}
+
+	// Query performance data
+	samples, err := perfManager.Query(ctx, []types.PerfQuerySpec{spec})
+	if err != nil {
+		return metricMap, fmt.Errorf("failed to query performance data: %w", err)
+	}
+
+	if len(samples) == 0 {
+		m.Logger().Debug("No samples returned from performance manager")
+		return metricMap, nil
+	}
+
+	results, err := perfManager.ToMetricSeries(ctx, samples)
+	if err != nil {
+		m.Logger().Errorf("failed to convert performance data to metric series: %v", err)
+	}
+
+	for _, result := range results[0].Value {
```

Comment:
we should make sure that `results` has at least one value like we have done for samples. Or it could be a potential panic.

Reply:
There is another check before it if there are zero samples for results then it will return from there itself and if there is any issue while converting it to series we will return with error.

Reply:
Got your point. However, looking at the definition of `ToMetricSeries`, I still see two potential issues with it
-  If the type assertion `s, ok := series[i].(*types.PerfEntityMetric)` fails, the function panics. This could lead to a crash unless you want this behavior. Should we return an error instead of panicking to handle unexpected input more gracefully?
- The line `v := s.Value[j].(*types.PerfMetricIntSeries)` assumes that all metrics in s.Value are of type *types.PerfMetricIntSeries. If that's not guaranteed (or there are other possible types), this will panic.

Both cases could potentially result in a crash. Unless we intend to abort the execution or expect this to happen, we should be handling it.

Reply:
I agree with Aman. The only case it is returning an error is when CounterInfoByKey fails. Even if it passes; it doesn't guarantee that the returned slice from `ToMetricSeries` will have at least 1.

So, we should ideally handle it here just to be on the safe side.

Reply:
You are right about panic.
We should probably handle it.
Either we handle panic or we need to manually convert to the series.
I can create new method as well because we are already doing `CounterInfoByName`. I can just reuse it.
And for other stuff we can return error instead of Panic.
What do you think?

Reply:
Adding recover to make sure code does not fail.

Reply:
got it.

Reply:
I checked this pretty quick; will request you to verify as well as I am busy in some other work:

Query: https://github.com/vmware/govmomi/blob/c1151f859ba649771c4ab60992953019904355fe/performance/manager.go#L268 

Query calls QueryPerf and that makes sure that the types are right: https://github.com/vmware/govmomi/blob/c1151f859ba649771c4ab60992953019904355fe/simulator/performance_manager.go#L219

Execept for length, I don't think we need to check for types.

Reply:
It can only iff https://github.com/vmware/govmomi/blob/c1151f859ba649771c4ab60992953019904355fe/simulator/performance_manager.go#L249 is done. But you are not setting the format in here:

```
spec := types.PerfQuerySpec{
		Entity:     hst.Reference(),
		MetricId:   metricIds,
		MaxSample:  1,
		IntervalId: refreshRate,
	}
```